### ISSUE TITLE
OAuth Redirect Authentication flow

### ISSUE THREAD

**Issue by andrewazores** [cryostatio/cryostat#717]
https://vertx.io/docs/vertx-auth-oauth2/java/

Vert.x supports OAuth redirect authentication flow. Cryostat could support this when using the OpenShiftAuthManager, or any other AuthManager implementation that delegates to an OAuth server. This way the user would not supply their Bearer token directly to Cryostat. Instead, the user would visit the Cryostat web-client and be redirected to the platform OAuth server, allowing the user to log in with whatever credentials that server is configured to require (OpenShift cluster username/password, LDAP, or some other SSO). After successful authentication the user would be redirected back to the web-client with their access token as a query parameter, which the web-client should be able to capture and store as usual.

**Comment by andrewazores** [comment]
I think the issue can remain open - there is still some work to do on the AuthPostHandler so that it can send the proper redirect to the client. This should obviously only happen when the AuthManager is one where that makes sense, so there will also need to be some work done on the AuthManager and its implementations to sort out how to handle that redirect flow.

**Comment by jan-law** [comment]
Here's what I've found about the `scope` parameter while using the service account as an OAuth client.

The OpenShift docs [here](https://docs.openshift.com/container-platform/4.9/authentication/using-service-accounts-as-oauth-client.html#service-accounts-as-oauth-clients_using-service-accounts-as-oauth-client) mention that service accounts acting as OAuth clients have a reduced set of scopes, meaning it can't request a `user:full` access to all API permissions. Attempting to request `scope=user:full` results in `error = access_denied & error_description = scope denied user:full`. If I leave the scope field empty, the OAuth server assumes I requested the scope `user:full` and denies access. 

Instead of specifying `user:full`, we could also request a scope with any role in the namespace. Looking at the roles from `oc get roles`, I tried requesting a scope with the roles `scope=role:cryostat-sample:<namespace>` and `scope=role:cryostat-operator-role:<namespace>` which both give me the same 401 error in the web client. 

Any ideas about how we could reuse the existing RBAC permissions to request a token with full access to the API?

```
HTTP Authorization Failure caused by KubernetesClientException: 
Failure executing: POST at: https://172.30.0.1/apis/authorization.k8s.io/v1/selfsubjectaccessreviews. 
Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. 
selfsubjectaccessreviews.authorization.k8s.io is forbidden: User "kube:admin" cannot create resource 
"selfsubjectaccessreviews" in API group "authorization.k8s.io" at the cluster scope: scopes [role:cryostat-sample:cryostat-
operator-system] prevent this action.
```

**Comment by andrewazores** [comment]
The operator's `config/rbac/role.yaml` and `config/rbac/cryostat_role.yaml`, which define the service account Roles (cryostat and the operator), both do include that `create selfsubjectaccessreviews` capability. The `scope` you supply does seem to map correctly to these Roles. But, you are not authenticating to OAuth as the machine service account, rather as the user account `kube:admin` - that makes sense, but that user isn't bound to the same Roles and in this case doesn't have any Role that gives it the `selfsubjectaccessreview` capability.

I don't know if the `scope` we provide needs to map to one of the `Roles` we have defined for service accounts to use. The logged in end user doesn't need to have an identical set of permissions as to what Cryostat's own service account does.

> Attempting to request scope=user:full results in error = access_denied & error_description = scope denied user:full. If I leave the scope field empty, the OAuth server assumes I requested the scope user:full and denies access.

When/where does this occur? Is this in the browser after you try to log in graphically, or is this happening at some time when Cryostat is trying to make an OAuth API request using its own service account token?

What is the interaction between OAuthClient and serviceaccount? In my mind the overall model I have is that the OAuthClient resource just represented the web-client/its redirect URI/the user interacting through it, so when the user is going through the redirect login flow they are communicating directly to the OAuth server and not through a service account. They provide their credentials and the server sends them back their OAuth token for whatever scope is required. Separately, Cryostat has its own service account and token, which it can use to communicate with the same OAuthServer as its own form of limited OAuthClient, and at times it does these communications while masquerading as the user after the user has supplied their token along with some attempt to perform an authenticated action. But at the initial login stage where the user is acquiring their token from the OAuth server the Cryostat service account hasn't been involved yet, right?

**Comment by jan-law** [comment]
> When/where does this occur? 

After I login from the OpenShift Container Platform login page, the OAuth server returns the redirect URI with an error description instead of a token: `https://cryostat-sample-oauth.apps.jalaw0.lab.upshift.rdu2.redhat.com/?error=access_denied&error_description=scope+denied+user%3Afull`

> What is the interaction between OAuthClient and serviceaccount? 

Based on the description from the [Implicit Grant Flow rfc](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2), I think the “client” is our service account, “user-agent” is the web browser, “authorization server” is the OpenShift OAuth server, “resource owner” is a human user that knows their `kube:admin` credentials, and “web-hosted client resource” is the cryostat-sample app. In part (A), it says “the client initiates the flow by directing the resource owner’s user-agent to the authorization endpoint. The client includes its client identifier”. I think this means that OAuthClients are the only entities that can make requests to the OAuth server. As long as the service account contains a redirect URI, the service account becomes the entity making requests to the OAuth server. 

When the `oauth/authorize` URL contains `client_id=<service account name>`, the OAuth server treats the service account as a valid OAuthClient and returns the HTML for the login page. If the OAuth server can’t find an OAuth client from the `client_id` parameter, it returns a blank HTML page and various `400` errors.

Visiting `oauth/authorize` with `client_id=<service account name>`, with a service account that does NOT contain a redirect uri returns
`{"error":"unauthorized_client","error_description":"The client is not authorized to request a token using this method."}`

Visiting `oauth/authorize` and omiting the `client_id` parameter returns
`{"error":"server_error","error_description":"The authorization server encountered an unexpected condition that prevented it from fulfilling the request."}`

**Comment by jan-law** [comment]
> The logged in end user doesn't need to have an identical set of permissions as to what Cryostat's own service account does.

I realized that the `role` specified in a role scope is a `clusterRole`, not the roles from `oc get roles`. I'll look into the `clusterRoles` that the operator already creates and see if there's one we can apply as a role scope

**Comment by andrewazores** [comment]
> > When/where does this occur?
> 
> After I login from the OpenShift Container Platform login page, the OAuth server returns the redirect URI with an error description instead of a token: `https://cryostat-sample-oauth.apps.jalaw0.lab.upshift.rdu2.redhat.com/?error=access_denied&error_description=scope+denied+user%3Afull`
> 
> > What is the interaction between OAuthClient and serviceaccount?
> 
> Based on the description from the [Implicit Grant Flow rfc](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2), I think the “client” is our service account, “user-agent” is the web browser, “authorization server” is the OpenShift OAuth server, “resource owner” is a human user that knows their `kube:admin` credentials, and “web-hosted client resource” is the cryostat-sample app. In part (A), it says “the client initiates the flow by directing the resource owner’s user-agent to the authorization endpoint. The client includes its client identifier”. I think this means that OAuthClients are the only entities that can make requests to the OAuth server. As long as the service account contains a redirect URI, the service account becomes the entity making requests to the OAuth server.
> 
> When the `oauth/authorize` URL contains `client_id=<service account name>`, the OAuth server treats the service account as a valid OAuthClient and returns the HTML for the login page. If the OAuth server can’t find an OAuth client from the `client_id` parameter, it returns a blank HTML page and various `400` errors.
> 
> Visiting `oauth/authorize` with `client_id=<service account name>`, with a service account that does NOT contain a redirect uri returns `{"error":"unauthorized_client","error_description":"The client is not authorized to request a token using this method."}`
> 
> Visiting `oauth/authorize` and omiting the `client_id` parameter returns `{"error":"server_error","error_description":"The authorization server encountered an unexpected condition that prevented it from fulfilling the request."}`

Thanks, this all makes sense and your interpretation of `client`, `user-agent`, etc. sounds dead-on.

> I realized that the role specified in a role scope is a clusterRole, not the roles from oc get roles.
Role Scope docs
I'll look into the clusterRoles that the operator already creates and see if there's one we can apply as a role scope

Sounds good - that would explain why setting the `scope` to a Role that should be able to do the `selfsubjectaccessreview`s wasn't working, I suppose.

If there are no existing cluster roles that meet our needs then I think we can discuss with @ebaron about adding a new `CryostatOAuth` ClusterRole. That probably makes sense to do regardless before we finish fleshing out and merging this feature, because we want to ensure that that ClusterRole only has the exact subset of permissions it really needs, and so repurposing some other ClusterRole is just asking for trouble - even if there is a role that happens to currently have the exact set of permissions by happenstance.

**Comment by ebaron** [comment]
I'm a bit confused about the role scope. This is so we get a token that has the permissions needed by the permissions API, right?

If I remember correctly, the token the user enters manually is only used to do a TokenReview and SelfSubjectAccessReview. In that case, would we be able to get away with just using the `user:check-access` scope? The TokenReview could be removed since the OAuth server should authenticate the user for us, and the check-access scope will allow us to do the SSAR.

**Comment by jan-law** [comment]
Here's what I know:
When we request a token from the OAuth server with a `clusterRole` as the role scope, the OAuth server will return a token that has the same permissions as the `clusterRole`. Then whenever the web-UI makes an API request to the backend with that same token, that token needs to have enough permissions for the backend to fulfill any API requests, including creating recordings and this `performTokenReview()` function I found below. 

https://github.com/cryostatio/cryostat/blob/d486871f35479b18c1a44360fe4e18697bee58ff/src/main/java/io/cryostat/net/OpenShiftAuthManager.java#L270
As of now, I've requested both TokenReview, SelfSubjectAccessReview, and `user:check-access` in the scope. If I remove the `TokenReview create` permission, the cryostat logs outputs this error below. Omitting either the SelfSubjectAccessReview or `user:check-access` scope also results in a similar exception.
```
INFO: Exception thrown
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://172.30.0.1/apis/authentication.k8s.io/v1/tokenreviews. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. tokenreviews.authentication.k8s.io is forbidden: User "system:serviceaccount:default:cryostat-sample" cannot create resource "tokenreviews" in API group "authentication.k8s.io" at the cluster scope.
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:639)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:576)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:543)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:504)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:292)
	at io.fabric8.kubernetes.client.dsl.internal.CreateOnlyResourceOperationsImpl.create(CreateOnlyResourceOperationsImpl.java:67)
	at io.cryostat.net.OpenShiftAuthManager.performTokenReview(OpenShiftAuthManager.java:317)
	at io.cryostat.net.OpenShiftAuthManager.reviewToken(OpenShiftAuthManager.java:195)
	at io.cryostat.net.OpenShiftAuthManager.validateToken(OpenShiftAuthManager.java:167)
	at io.cryostat.net.OpenShiftAuthManager.validateHttpHeader(OpenShiftAuthManager.java:268)
	at io.cryostat.net.OpenShiftAuthManager.sendLoginRedirectIfRequired(OpenShiftAuthManager.java:142)
	at io.cryostat.net.web.http.api.v2.AuthPostHandler.handle(AuthPostHandler.java:104)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:117)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:69)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

I'm not sure which exact permissions the backend needs to complete all of the web-UI actions. I essentially started with an empty `clusterRole` and added one permission at a time whenever the web-UI gave me a similar `accessDenied` exception to the one below:
```
Request failed (401 Unauthorized)
HTTP Authorization Failure caused by OpenShiftAuthManager.PermissionDeniedException: Requesting client in namespace
 "default" cannot patch flightrecorders.operator.cryostat.io: scopes [user:check-access
 role:cryostat-operator-cryostat:default] prevent this action
```

**Comment by andrewazores** [comment]
> I'm not sure which exact permissions the backend needs to complete all of the web-UI actions.

This is something we should probably make easier to determine. The permissions required for each action are defined by the API handlers themselves:

https://github.com/cryostatio/cryostat/blob/d486871f35479b18c1a44360fe4e18697bee58ff/src/main/java/io/cryostat/net/web/http/RequestHandler.java#L74

That set of `ResourceAction`s gets mapped to platform-specific actions and resources by the `AuthManager` implementation before making the actual authz query to the backing auth server (OAuth in this case):

https://github.com/cryostatio/cryostat/blob/d486871f35479b18c1a44360fe4e18697bee58ff/src/main/java/io/cryostat/net/OpenShiftAuthManager.java#L166 (the two `map` calls for `getResource()` and `getVerb()`)

We could generate a manifest of the required Cryostat application-level permissions easily enough by ex. implementing a new `RequestHandler` that has injected a `Lazy<Set<RequestHandler>>` so that it can get a reference to all of the handlers. Then, call `resourceActions()` on each of these handlers and perform set union on the results. This gives you the entire set of required permissions that are actually requestable through the API.

From there, if we hoist those `map()` calls from the `OpenShiftAuthManager` into the `AuthManager` interface so that all implementations have something similar, then we can map the set of Cryostat application permissions into actual platform-specific permissions. Apply another pass of filtering out empty permissions, since some Cryostat application permissions may translate into nothing (implicitly granted) and you have a nice platform-specific manifest of all permissions required for a user or service account to have access to all features.

**Comment by ebaron** [comment]
> Here's what I know: When we request a token from the OAuth server with a `clusterRole` as the role scope, the OAuth server will return a token that has the same permissions as the `clusterRole`. Then whenever the web-UI makes an API request to the backend with that same token, that token needs to have enough permissions for the backend to fulfill any API requests, including creating recordings and this `performTokenReview()` function I found below.
> 
> https://github.com/cryostatio/cryostat/blob/d486871f35479b18c1a44360fe4e18697bee58ff/src/main/java/io/cryostat/net/OpenShiftAuthManager.java#L270
> 
> As of now, I've requested both TokenReview, SelfSubjectAccessReview, and `user:check-access` in the scope. If I remove the `TokenReview create` permission, the cryostat logs outputs this error below. Omitting either the SelfSubjectAccessReview or `user:check-access` scope also results in a similar exception.
> 
> ```
> INFO: Exception thrown
> io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://172.30.0.1/apis/authentication.k8s.io/v1/tokenreviews. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. tokenreviews.authentication.k8s.io is forbidden: User "system:serviceaccount:default:cryostat-sample" cannot create resource "tokenreviews" in API group "authentication.k8s.io" at the cluster scope.
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:639)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:576)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:543)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:504)
> 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:292)
> 	at io.fabric8.kubernetes.client.dsl.internal.CreateOnlyResourceOperationsImpl.create(CreateOnlyResourceOperationsImpl.java:67)
> 	at io.cryostat.net.OpenShiftAuthManager.performTokenReview(OpenShiftAuthManager.java:317)
> 	at io.cryostat.net.OpenShiftAuthManager.reviewToken(OpenShiftAuthManager.java:195)
> 	at io.cryostat.net.OpenShiftAuthManager.validateToken(OpenShiftAuthManager.java:167)
> 	at io.cryostat.net.OpenShiftAuthManager.validateHttpHeader(OpenShiftAuthManager.java:268)
> 	at io.cryostat.net.OpenShiftAuthManager.sendLoginRedirectIfRequired(OpenShiftAuthManager.java:142)
> 	at io.cryostat.net.web.http.api.v2.AuthPostHandler.handle(AuthPostHandler.java:104)
> 	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:117)
> 	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:69)
> 	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
> 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
> 	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> ```
> 
> I'm not sure which exact permissions the backend needs to complete all of the web-UI actions. I essentially started with an empty `clusterRole` and added one permission at a time whenever the web-UI gave me a similar `permissionDenied` exception to the one below:
> 
> ```
> Request failed (401 Unauthorized)
> HTTP Authorization Failure caused by OpenShiftAuthManager.PermissionDeniedException: Requesting client in namespace
>  "default" cannot patch flightrecorders.operator.cryostat.io: scopes [user:check-access
>  role:cryostat-operator-cryostat:default] prevent this action
> ```

Thanks for the explanation Janelle! The OpenShift documentation about the scopes could be a bit more precise. FWIW, I seem to have found the code where the scopes are translated into RBAC objects, so we can see exactly what permissions they give: https://github.com/openshift/apiserver-library-go/blob/5cdb70a1e65b6bcabb3b897e38287ed2c8ed77d1/pkg/authorization/scope/converter.go

It looks like the `user:access` scope doesn't let you find out what the user can access, but what the scoped token can access. If neither, `user:full` or `role:<whatever>` are provided, then it's pretty much useless. Seems a bit odd to me, but now I understand why we need the cluster role.

Interestingly `user:info` grants access to `user.openshift.io/v1/users/~` which returns information on the user: https://docs.openshift.com/container-platform/4.9/rest_api/user_and_group_apis/user-user-openshift-io-v1.html. We could likely replace the TokenReview with this. This would cover all the permissions in https://github.com/cryostatio/cryostat-operator/blob/e524857960ffdfe44eaa360e9fd215d14b968953/config/rbac/cryostat_role.yaml, which would mean the operator doesn't need to create a ClusterRoleBinding for each Cryostat deployment. The ClusterRole could then just be used for the purpose of the role scope.

**Comment by jan-law** [comment]
> Then, call resourceActions() on each of these handlers and perform set union on the results. This gives you the entire set of required permissions that are actually requestable through the API.

I made a handler on [jan-law:list-permissions](https://github.com/jan-law/cryostat/commit/cd8335aaaacdc54d2b194c8034062f38e303cc37) that outputs this list of permissions: https://gist.github.com/jan-law/8bedb15a7027d2697191ea8b01dfd856

Some of the `ResourceType`s are easy to match with their corresponding `apiGroup`, eg `ResourceType RECORDINGS` and `operator.cryostat.io.recordings`. `TARGET` is most likely for pods or deployments.

How do the rest of the `ResourceTypes` correspond to the permissions in the clusterRole? And why don't these `ResourceActions` refer to TokenReviews or SelfSubjectAccessReviews?

**Comment by andrewazores** [comment]
> How do the rest of the ResourceTypes correspond to the permissions in the clusterRole

They don't necessarily - not everything that I defined as a resource from Cryostat's POV has been mapped to something in OpenShift RBAC. Not yet, anyway, and for some of them maybe not ever. There is no CRD for managing Credentials, for example, and so there's no RBAC mapping there.

> And why don't these ResourceActions refer to TokenReviews or SelfSubjectAccessReviews?

TokenReviews and SelfSubjectAccessReviews are OpenShift-specific auth implementation details, so they don't show up in Cryostat's application-level model of resources/actions. Maybe we should have a `ResourceType` for something like `USER` though - `AuthPostHandler` currently has `ResourceActions.NONE`.

**Comment by ebaron** [comment]
> > Then, call resourceActions() on each of these handlers and perform set union on the results. This gives you the entire set of required permissions that are actually requestable through the API.
> 
> I made a handler on [jan-law:list-permissions](https://github.com/jan-law/cryostat/commit/cd8335aaaacdc54d2b194c8034062f38e303cc37) that outputs this list of permissions: https://gist.github.com/jan-law/8bedb15a7027d2697191ea8b01dfd856
> 
> Some of the `ResourceType`s are easy to match with their corresponding `apiGroup`, eg `ResourceType RECORDINGS` and `operator.cryostat.io.recordings`. `TARGET` is most likely for pods or deployments.

Looks good! If you cross-reference that output with the mapping here, you'll have a complete list: https://github.com/cryostatio/cryostat/blob/05955511d9b5147ad93a5f844d6e408db68dbe98/src/main/java/io/cryostat/net/OpenShiftAuthManager.java#L312-L345

It should be some subset of CRUD on:
* pods
* deployments
* cryostats
* flightrecorders
* recordings

**Comment by jan-law** [comment]
Thanks! Here's the output:
```
        "[CRYOSTATS, PODS, DEPLOYMENTS]: create",
        "[CRYOSTATS]: create",
        "[CRYOSTATS]: delete",
        "[FLIGHTRECORDERS]: create",
        "[FLIGHTRECORDERS]: delete",
        "[FLIGHTRECORDERS]: get",
        "[FLIGHTRECORDERS]: patch",
        "[PERMISSION_NOT_REQUIRED]: create",
        "[PERMISSION_NOT_REQUIRED]: delete",
        "[PERMISSION_NOT_REQUIRED]: get",
        "[RECORDINGS]: create",
        "[RECORDINGS]: delete",
        "[RECORDINGS]: get",
        "[RECORDINGS]: patch"
```
When I ran Cryostat on OpenShift with the same clusterRole permissions as above, with the role scope set to `user:info role:cryostat-operator-cryostat`, I got `PermissionDeniedExceptions` for the following permissions, so I added these back into `cryostat_role.yaml`:
- `get deployments`
- `get pods`
- `get cryostats`
- `create TokenReviews`

**Comment by ebaron** [comment]
Ah, the first two are probably from the discovery/tree API. It's capable of making the following `get` requests [1]:
- statefulsets
- daemonsets
- deployments
- replicasets
- replicationcontrollers
- pods
- endpoints

I'm not sure where the `get cryostats` comes from though.

[1] https://github.com/cryostatio/cryostat/blob/main/src/main/java/io/cryostat/platform/internal/KubeApiPlatformClient.java#L366-L386

**Comment by ebaron** [comment]
The `get cryostats` comes from the Messaging Server. It checks that you have `get` permissions for all of the `ResourceTypes`.

https://github.com/cryostatio/cryostat/blob/05955511d9b5147ad93a5f844d6e408db68dbe98/src/main/java/io/cryostat/messaging/MessagingServer.java#L148-L156

**Comment by jan-law** [comment]
As of now, access tokens expire in 24 hours, which means if you click "Logout", any backend queries to the OAuth server will return the existing token instead of redirecting to the OpenShift Container Platform login page. Would you prefer if I made a separate PR to add a logout capability or add it to this one?

**Comment by andrewazores** [comment]
It might be easier to review as a separate follow-up PR.

**Comment by ebaron** [comment]
I think once this is finished, we should verify the complete workflow with both kubeadmin and regular users. There seems to be at least some difference in how authentication works between them:
https://github.com/openshift/console/blob/22c6951efe7c4bca87f3f934063b9f4dcb0a4058/frontend/public/module/auth.js#L70-L80
